### DISCUSSION TITLE
introspection

### DISCUSSION THREAD

**Discussion by kmerfeld** [mattwparas/steel#132] [Q&A] [answered]
Hello, Is there a a way to inspect a function? I found `inspect-bytecode` and the generated `{$fn_name}__doc__`, but is there a way to see the lisp code for an existing function?

**Comment by mattwparas** [comment]
Hey - so do you want to see the original code? Or the fully macro expanded code?

if you run `steel ast <file-name>` - that should spit out the fully expanded ast (similarly you can run `steel bytecode <filename>` and it will do the same, but for the bytecode)

Unfortunately I don't have a way yet to do something like `(inspect-ast <function>)` yet - is that something that would be valuable? I can very easily do it for the non expanded ast, but its a bit more difficult for the fully expanded ast

**Reply by kmerfeld** [reply]
I guess I'm looking for a way to view the non expanded ast in the repl.
Ideally this would work for SteelVal::Closure, SteelVal::FuncV, and SteelVal::BuiltIn, but I'm not sure how that would work

**Reply by mattwparas** [reply]
I don't think `FuncV` or `BuiltIn` would make a whole lot of sense - those are primitive rust functions so showing the source code (while not impossible) seems a bit unnecessary.

For showing the original ast - is this just to help refresh your memory on what it does? Just trying to understand the motivation behind having it accessible via the repl. It would bloat the runtime a bit to include it, however I could use a flag on whether to keep it around or not.

**Reply by kmerfeld** [reply]
If all FuncV and BuiltIn had docs it wouldn't be a big issue. right now steel doesn't have great documentation, so I find I need to grep the steel repo constantly to find the function I need.
It would be really nice to be able to explore the language more from the repl

**Reply by mattwparas** [reply, marked as answer]
You're right, the docs aren't great. However here are the steps I'm taking to improve on that:

* Online docs can be found [here](https://mattwparas.github.io/steel/book/reference/builtins.html). Functions which are documented will show up here after running a doc generation script.
* You can use the LSP which is wired up to find docs as well - in the repo there is an lsp crate with instructions how to use it in the [README](https://github.com/mattwparas/steel/tree/master/crates/steel-language-server)

So with respect to that - I am trying to improve on it and slowly working towards it. If at any point you become comfortable enough with the codebase and would like to contribute docs, they are always welcome

For exploring via the repl, I believe things should be wired up so that once docs exist, the repl should have access. I can also see the overhead for exploring the ast from the repl as well. In theory it should be pretty straightforward since the macro which creates the doc comment variable is in the standard library, and we have access to the ast there. However, only functions that have doc comments on them would have their ast bound.

**Reply by mattwparas** [reply]
On that note - I have a PR open now that will add docs for immutable vectors and hashsets, and I'll slowly chunk through the existing functions to make sure those also have docs
